src/cascadia/TerminalConnection/AzureConnection.cpp

```diff
@@ -0,0 +1,478 @@
+        {
+            throw "Time out.";
+        }
+        return responseJson;
+    }
+
+    // Helper function to acquire the user's Azure tenants
+    json::value AzureConnection::_getTenants()
+    {
+        //Initialize the client
+        http_client tenantClient(_resourceUri);
+
+        //Initialize the request
+        http_request tenantRequest(L"GET");
+        tenantRequest.set_request_uri(L"tenants?api-version=2018-01-01");
+        tenantRequest.headers().add(L"Accept", L"application/json");
+        tenantRequest.headers().add(L"Authorization", L"Bearer " + _accessToken);
+        tenantRequest.headers().add(L"User-Agent", L"Terminal");
+
+        //Send the request and return the response as a json value
+        return _requestHelper(tenantClient, tenantRequest);
+    }
+
+    json::value AzureConnection::_refreshTokens()
+    {
+        //Initialize the client
+        http_client refreshClient(_loginUri);
+
+        //Initialize the request
+        http_request refreshRequest(L"POST");
+        refreshRequest.set_request_uri(_tenantID + L"/oauth2/token");
+        utility::string_t body = L"client_id=" + AzureClientID + L"&resource=" + _wantedResource + L"&grant_type=refresh_token" + L"&refresh_token=" + _refreshToken;
+        refreshRequest.set_body(body, L"application/x-www-form-urlencoded");
+        refreshRequest.headers().add(L"User-Agent", L"Terminal");
+
+        //Send the request and return the response as a json value
+        return _requestHelper(refreshClient, refreshRequest);
+    }
+
+    // Helper function to get the user's cloud shell settings
+    json::value AzureConnection::_getCloudShellUserSettings()
+    {
+        //Initialize client
+        http_client settingsClient(_resourceUri);
+
+        //Initialize request
+        http_request settingsRequest(L"GET");
+        settingsRequest.set_request_uri(L"providers/Microsoft.Portal/userSettings/cloudconsole?api-version=2018-10-01");
+        settingsRequest.headers().add(L"Accept", L"application/json");
+        settingsRequest.headers().add(L"Content-Type", L"application/json");
+        settingsRequest.headers().add(L"Authorization", L"Bearer " + _accessToken);
+        settingsRequest.headers().add(L"User-Agent", L"Terminal");
+
+        return _requestHelper(settingsClient, settingsRequest);
+    }
+
+    utility::string_t AzureConnection::_getCloudShell()
+    {
+        //Initialize client
+        http_client cloudShellClient(_resourceUri);
+
+        //Initialize request
+        http_request shellRequest(L"PUT");
+        shellRequest.set_request_uri(L"providers/Microsoft.Portal/consoles/default?api-version=2018-10-01");
+        shellRequest.headers().add(L"Accept", L"application/json");
+        shellRequest.headers().add(L"Content-Type", L"application/json");
+        shellRequest.headers().add(L"Authorization", L"Bearer " + _accessToken);
+        shellRequest.headers().add(L"User-Agent", L"Terminal");

```

Comment:
It seems like many of these requests use a common set of headers - maybe we want to create a helper method to initialize a common `http_request` object?

Reply:
I will write a helper to add the headers to a given http_request object, because the http_requests themselves vary (between GET, POST and PUT)